lib/logging/third_party_transaction.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module Logging
+  module ThirdPartyTransaction
+    module MethodWrapper
+      # @method_names
+      #   - Array (of /or single) method name(s). e.g. ['method_1', :method_2]
+      # @additional_logs
+      #   - Will add to and / or overwrite the default data with usecase specific
+      #     log data
+      def wrap_with_logging(*method_names, additional_logs: {})
+        # including the instance method helpers inside this method makes them
+        # available on the instance with access to the current scope, (e.g.
+        # current_user at the controller level)
+        include(ScopedInstanceMethods)
+
+        proxy = Module.new do
+          method_names.each do |method_name|
+            # define patchable method(s) with the same name inside of a proxy module.
+            define_method(method_name) do |*args, &block|
+              log_3pi_begin(method_name, additional_logs, *args)
+              # delegate to the original behavior
+              result = super(*args, &block)
+              log_3pi_complete(method_name, additional_logs, *args)
+              result
+            end
+          end
+        end
+        # prepend the proxy module, allowing both our wrapped version to run
+        # logging as well as delegating to the original.
+        prepend(proxy)
+      end
+    end
+
+    # these will be included after instance instantiation, making them available
+    # to the instance and retaining their scope.
+    module ScopedInstanceMethods
+      def log_3pi_begin(method_name, additional_logs = {}, args = [])
+        @start_time = Time.current
+
+        log = {
+          process_id: Process.pid,
+          user_uuid: try(:current_user).try(:account_uuid),
+          action: 'Begin interaction with 3rd party API',
+          wrapped_method: "#{self.class}##{method_name}",
+          start_time: @start_time,
+          passed_args: args.to_s
+        }.merge(additional_logs)
+
+        Rails.logger.info(log)
+      rescue => e
+        Rails.logger.error(e)
+      end
+
+      def log_3pi_complete(method_name, additional_logs = {}, args = [])
+        now = Time.current
```

Comment:
(Blocking pending peace of mind answer) I'd double check the methods you are using for `Time` objects are consistent with other time-based logging elsewhere because of potential confusion related to timezones.

My understanding is that what are you using here, `Time.current`, which retrieves the time in the application's configured timezone, is the best method to use universally.

However, if you look at both our [application config](https://github.com/department-of-veterans-affairs/vets-api/blob/master/config/application.rb) and [production env config](https://github.com/department-of-veterans-affairs/vets-api/blob/master/config/environments/production.rb) it doesn't appear we configure the time zone for the application, and in that case I think it defaults to the time zone of whatever system the application is running on.

That may be ok but it's just something I thought was worth calling out in case you missed it.

Reply:
[PENDING ACTION] good idea, will do

Reply:
Confirmed that this same format is used in at least a high number of other cases.  here is an example from our [Webhooks::SchedulerJob](https://github.com/department-of-veterans-affairs/vets-api/blob/615c41468e1bbdb247664683aed4a5b82691ac74/app/workers/webhooks/scheduler_job.rb#L11).  I checked to see what `processing_time` was (you can see it set in the retry [a few lines down here](https://github.com/department-of-veterans-affairs/vets-api/blob/615c41468e1bbdb247664683aed4a5b82691ac74/app/workers/webhooks/scheduler_job.rb#L23)

Checking the class of `5.minutes.from_now` against `Time.current` reveals a match.  For sanity, here is what that output will look like in a log
`Fri, 28 Jul 2023 19:18:32.825530000 UTC +00:00`
I *cannot* confirm this is our only time format being logged, but at least we have a precedent, which feels good IMHO.

<img width="665" alt="Screen Shot 2023-07-28 at 3 14 58 PM" src="https://github.com/department-of-veterans-affairs/vets-api/assets/15328092/16de07b8-6dbf-4252-802b-76751fe21035">